transformer_engine/pytorch/cpp_extensions/normalization.py

```diff
@@ -13,7 +13,37 @@
            'layernorm_fwd_inf',
            'rmsnorm_fwd_fp8',
            'rmsnorm_fwd_fp8_inf',
-           'rmsnorm_fwd_inf']
+           'rmsnorm_fwd_inf',
+           'get_norm_workspace_and_barrier',
+           'set_norm_workspace_and_barrier']
+
+
+_norm_scratch_spaces = {}
+
+
+def get_norm_key(inp: torch.Tensor, weight: torch.Tensor, fp8: bool) -> str:
+    """Get unique key for workspace/barrier config."""
+    return f"{inp.shape}_{inp.dtype}_{weight.shape}_{weight.dtype}_{fp8}"
+
+
+def get_norm_workspace_and_barrier(
+    inp: torch.Tensor,
+    weight: torch.Tensor,
+    fp8: bool,
+) -> Tuple[Union[torch.Tensor, None], Union[torch.Tensor, None]]:
+    """Get workspace and barrier for config."""
+    key = get_norm_key(inp, weight, fp8)
+    return key, _norm_scratch_spaces.get(key, (None, None))
+
+
+def set_norm_workspace_and_barrier(
```

Comment:
Why do we need to cache the layernorm workspace and barrier buffers? They're allocated and deallocated within the graph capture, so I'd expect they'd be handled correctly by the memory allocator. Calculating their dimensions only requires CPU logic too, so GPU synchronization isn't an issue.

Reply:
This is needed for pipeline parallelism where we have multiple cuda graphs per module and the order of fwd/bwd execution is non-trivial (not fbfbfb...). The intermediate buffers are freed during after capture which result in memory corruptions of subsequent cuda graphs' memory pools. I will explain in greater depth when I update the PR description.

Reply:
Removed